DFGM/dfgm_subsystem.py

```diff
@@ -0,0 +1,221 @@
+
+Usage: dfgm_subsystem.py non-default_port_num
+
+Copyright 2023 [Daniel Sacro]. Licensed under the Apache License, Version 2.0
+"""
+
+import sys
+import socket
+import time
+from struct import pack
+
+DEFAULT_HOST = '127.0.0.1'
+DEFAULT_PORT = 1802
+
+# Format/order of housekeeping data
+house_keeping_data = {
+    "Core Voltage": 5000, # HK 0 (mV)
+    "Sensor Temperature": 25, # HK 1 (deg C)
+    "Reference Temperature": 25, # HK 2 (deg C)
+    "Board Temperature": 25, # HK 3 (deg C)
+    "Positive Rail Voltage": 5000, # HK 4 (mV)
+    "Input Voltage": 5000, # HK 5 (mV)
+    "Reference Voltage": 5000, # HK 6 (mV)
+    "Input Current": 1000, # HK 7 (mA)
+    "Reserved 1": 0, # HK 8 (Unused)
+    "Reserved 2": 0, # HK 9 (Unused)
+    "Reserved 3": 0, # HK 10 (Unused)
+    "Reserved 4": 0, # HK 11 (Unused)
+}
+
+# Format of a raw magnetic field data sample to be processed by the OBC
+magnetic_field_tuple = {
+    "x_DAC": 1, 
+    "x_ADC": 1,
+    "y_DAC": 2,
+    "y_ADC": 2,
+    "z_DAC": 3,
+    "z_ADC": 3
+}
+
+# There are 100 samples in each packet from the DFGM
+magnetic_field_data = [magnetic_field_tuple] * 100
+
+# Format of the complete DFGM data packet
+default_packet = {
+    "DLE": 0x10, # Data Link Escape
+    "STX": 0x02, # Start of Text
+    "PID": 1, # Packet ID
+    "Packet Type": 1, # Type of data inside packet
+    "Packet Length": 1248, # In bytes
+    "FS": 100, # Sampling Frequency
+    "PPS Offset": 1, # "U32 offset in ticks from last PPS edge"
+    "HK_data": house_keeping_data,
+    "mag_data": magnetic_field_data, 
+    "Board ID": 1,
+    "Sensor ID": 1,
+    "Reserved 1": 55, # Reserved 1-5 are unused; reserved for any future uses
+    "Reserved 2": 55,
+    "Reserved 3": 55,
+    "Reserved 4": 55,
+    "Reserved 5": 55,
+    "ETX": 3, # End of Text
+    "CRC": 0 # Packet info
+}
+
+class DFGMSimulator:
+    '''Simulates the DFGM board's functionality'''
+
+    def __init__(self, client_socket):
+        self.client_socket = client_socket
+        self.is_first_packet = True
+        self.packet = None
+        self.house_keeping_bytes = None
+        self.magnetic_field_bytes = None
+        self.packet_bytes = bytearray(b'')
+
+    def start(self):
+        '''Simulates the DFGM board's ON state'''
+        while True:
+            try:
+                if self.is_first_packet:
+                    self.generate_packet()
+                    self.is_first_packet = False
+                else:
+                    self.update_packet()
+                self.format_packet()
+                self.send_packet()
+                self.print_packet()
+                time.sleep(1) # Force packet to send every 1 second
+            except BrokenPipeError:
+                print("Client disconnected abruptly")
+                break
+        self.client_socket.close()
+
+    def generate_packet(self):
+        '''Generates a new data packet'''
+        self.packet = default_packet
+
+    def update_packet(self):
```

Comment:
This is nit picky but would we be able to use random numbers or something of the like to better modify the fake voltage / current values?

Reply:
Yeah that's totally possible. I was debating on whether or not to use random numbers to change the housekeeping values before, so I can get that done